pynbody/filt.py

```diff
@@ -100,36 +100,50 @@ class Sphere(Filter):
     def __init__(self, radius, cen=(0, 0, 0)):
         self._descriptor = "sphere"
         self.cen = np.asarray(cen)
-        if self.cen.shape != (3,):
-            raise ValueError, "Centre must be length 3 array"
+#        if self.cen.shape != (3,):
+#            raise ValueError, "Centre must be length 3 array"
 
-        if isinstance(radius, str):
-            radius = units.Unit(radius)
+        if not isinstance(radius, list) : 
+            radius = [radius]
+
+        if isinstance(radius[0], str):
+            radius = map(units.Unit,radius)
+
+        # check if many different centers were passed in
+            if isinstance(cen, list) or isinstance(cen, np.ndarray) : 
+                if len(cen) != len(radius) : 
+                    radius = radius * len(cen)
 
         self.radius = radius
 
+
     def __call__(self, sim):
         radius = self.radius
 
         with sim.immediate_mode:
             pos = sim['pos']
 
-        if units.is_unit_like(radius):
-            radius = float(radius.in_units(pos.units,
-                                           **pos.conversion_context()))
+        if units.is_unit_like(radius[0]):
+            radius = np.asarray([r.in_units(pos.units, **pos.conversion_context()) for r in radius], dtype='float')
+
+        else : 
+            radius = np.asarray(radius, dtype='float')
 
         cen = self.cen
         if units.has_units(cen):
             cen = cen.in_units(pos.units)
+
+        if len(cen.shape) == 1 : 
+            cen = np.expand_dims(cen,axis=0)
+
         return util._sphere_selection(np.asarray(pos),np.asarray(cen,dtype=pos.dtype),radius)
 
 
     def __repr__(self):
-        if units.is_unit(self.radius):
-
-            return "Sphere('%s', %s)" % (str(self.radius), repr(self.cen))
+        if units.is_unit(self.radius[0]):
+            return " & ".join(["Sphere('%s', %s)" % (str(radius), repr(cen)) for radius, cen in zip(self.radius, self.cen)])
         else:
-            return "Sphere(%.2e, %s)" % (self.radius, repr(self.cen))
+            return " & ".join(["Sphere(%.2e, %s)" % (radius, repr(cen)) for radius, cen in zip(self.radius, self.cen)])
```

Comment:
Yes the notation is wrong -- the entire selection is negated. So the filter
originally *includes* all particles in the spheres but if you ~ it then
it's the inverse. It's an OR initially even though it says it's an AND...
Needs to be fixed
On Feb 6, 2015 7:04 PM, "salomanders" <notifications@github.com> wrote:

> In pynbody/filt.py
> <https://github.com/pynbody/pynbody/pull/236#discussion_r24258745>:
>
> >          else:
> > -            return "Sphere(%.2e, %s)" % (self.radius, repr(self.cen))
> > +            return " & ".join(["Sphere(%.2e, %s)" % (radius, repr(cen)) for radius, cen in zip(self.radius, self.cen)])
>
> I want to do a ~filter (cut out a sphere of particles to not include in
> the snapshot), and when it strings them together it only gives the ~ to the
> first sphere.
>
> filt = ~pynbody.filt.Sphere('0.1 kpc', st['pos'][0:4])
> ~Sphere('1.00e-01 kpc', array([ 910.50829655, 1563.14880045,
> 1451.38020969])) & Sphere('1.00e-01 kpc', array([ 913.86336456,
> 1566.48238759, 1451.69728888])) & Sphere('1.00e-01 kpc', array([
> 913.35059333, 1566.05790408, 1451.9326327 ])) & Sphere('1.00e-01 kpc',
> array([ 909.26119045, 1565.22082582, 1449.37438321]))
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pynbody/pynbody/pull/236/files#r24258745>.
>